### ISSUE TITLE
Boolean inputs do not work in `if:`

### ISSUE THREAD

**Issue by MetRonnie** [actions/runner#1483]
**Describe the bug**

Boolean type inputs do not work as expected in a step's `if` condition.

**To Reproduce**

Run this workflow using the workflow dispatch, with the foo input checkbox **not** ticked:
```yaml
name: Thingy

on:
  workflow_dispatch:
    inputs:
      foo:
        description: Whether to foo
        type: boolean
        required: false
        default: false

jobs:
  thingy:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    env:
      FOO: ${{ github.event.inputs.foo }}
    steps:
      - name: Check value
        run: |
          echo "The value is: $FOO"

      - name: Run if foo
        if: ${{ github.event.inputs.foo }}
        run: |
          echo "Foo!"

      - name: Don't run if foo
        if: ${{ ! github.event.inputs.foo }}
        run: |
          echo "Foo!"
```

**Expected behavior**

When the input is `false`, the `Run if foo` step should be skipped , and the `Don't run if foo` step should run.

## Runner Version and Platform
Version of your runner? `2.284.0`

OS of the machine running the runner? `ubuntu-latest (20.04.3)`

## What's not working?

The value of `github.event.inputs.foo` is `false`, yet the `Run if foo` step runs and the `Don't run if foo` step is skipped.

![image](https://user-images.githubusercontent.com/61982285/141504188-8702d06b-38c6-437f-b7f8-073e308c9e14.png)

## Additional context

It is the same for a string input type or boolean input type (https://github.blog/changelog/2021-11-10-github-actions-input-types-for-manual-workflows/).

Only `if: ${{ github.event.inputs.foo == 'true' }}` works.

This goes against what the documentation says at https://docs.github.com/en/actions/learn-github-actions/expressions#about-expressions

**Comment by ChristopherHX** [comment]
I guess this is a github service bug, sadly broken by design :(.

the value of the workflow_dispatch payload is `'false'`, but both of us are expecting a boolean `false` and not a string.
Why does this happen? `'false'` is a non empty string, which is truthy.
```
${{ github.event.inputs.foo }}
${{ 'false' }}
${{ true }}
```
However this might align to previous `workflow_dispatch` behavior, all inputs are strings like before.

This is super inconsistent with reusable workflows `workflow_call` ( beta ), because `type: boolean` does mean boolean and not a string like in `workflow_dispatch`.
```yaml
name: called
on:
  workflow_call:
    inputs:
      x:
        type: boolean

jobs:
  test1:
    runs-on: ubuntu-latest
    steps:
    - run: echo '${{inputs.x}}'
      if: ${{inputs.x}}
```
```yaml
name: caller
on:
  push:
jobs:
  test2:
    uses: <owner>/<repo>/.github/workflows/called.yml@<ref>
    with:
      x: false
```
![image](https://user-images.githubusercontent.com/44845461/141848557-14aebeea-654b-4f9d-8fef-87442ef77cf4.png)

**Comment by solarmosaic-kflorence** [comment]
I have run into this issue as well, it is tricky to identify and debug. I have just been ensuring I always cast potential boolean values to a string, and always match on the string value.

**Comment by TingluoHuang** [comment]
@ericsciple do you have any thoughts on this?

**Comment by piotrekkr** [comment]
Seems like input field type in workflow is only for GitHub to know what UI controls should it use and all inputs are strings... So it seems we should check it like `if: ${{ inputs.x == 'true' }}`

**Comment by solarmosaic-kflorence** [comment]
@piotrekkr it is the same for outputs, for example:

```yaml
outputs:
  foo:
    description: "Returns true or false"
    value: ${{ steps.foo.outputs.bar == 'true' }}
```

The `foo` output would be type boolean, like `false`, but matching on `outputs.foo == false` doesn't seem to work. So instead I do:

```yaml
outputs:
  foo:
    description: "Returns true or false"
    value: "${{ steps.foo.outputs.bar == 'true' }}"
```

And then match on `outputs.foo == 'true'` instead, which works every time.

**Comment by piotrekkr** [comment]
@solarmosaic-kflorence Yeah they should clearly document this behavior it can be really confusing. Maybe there is some documentation mentioning this but I did not stumbled upon this for a year now when working with GitHub...

**Comment by scottmmjackson** [comment]
This is actually really bad because there's no way to write a workflow such that if this bug is fixed your workflow continues to operate as normal. I can't figure out an expression that returns true only if a variable is not the string `'false'` and also not the boolean `false`, because any boolean test I do will be coercing the argument and returning `true` absent a literal empty string argument.

**Comment by ChristopherHX** [comment]
> I can't figure out an expression that returns true only if a variable is not the string `'false'` and also not the boolean `false`

Have you tried `${{fromJSON(github.event.inputs.foo)}}` or `${{fromJSON(github.event.inputs.foo)}}`? I'm not shure if this 100% does what you expect..
The trick `fromJSON` expects a string value, if it receives a boolean it is converted to string `false` => `'false'`, `true` => `'true'`. Then `fromJSON` translates the boolean string back to a real boolean.

You can also cast the input to a string with `${{format('{0}', github.event.inputs.foo)}}` then you always have the boolean as string. Then you could use `${{format('{0}', github.event.inputs.foo) != 'false'}}`, this might even work with malformed inputs which aren't valid json.

**Comment by piotrekkr** [comment]
> This is actually really bad because there's no way to write a workflow such that if this bug is fixed your workflow continues to operate as normal. I can't figure out an expression that returns true only if a variable is not the string `'false'` and also not the boolean `false`, because any boolean test I do will be coercing the argument and returning `true` absent a literal empty string argument.

@scottmmjackson Seems like something like this could work I think (still seems quite ugly):
```
${{ github.event.inputs.foo && github.event.inputs.foo != 'false' }}
```
should give
```
true    && true    != 'false'   => true
'true'  && 'true'  != 'false'   => true
false   && false   != 'false'   => false
'false' && 'false' != 'false'   => false
```

**Comment by kagahd** [comment]
I have a similar case by [reusing workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) where I want let the caller of the reused workflow decide about `continue-on-error`

I found the following workaround:

```
on:
  workflow_call:
    inputs:
      continue-on-error:
        type: boolean
        required: false
        default: false

jobs:
  my-job:
    name: 'my job'
    runs-on: ubuntu-latest
    env:
      CONTINUE_ON_ERROR: ${{inputs.continue-on-error}}
    steps:
      - uses: actions/checkout@v2
      - name: run foo
        shell: bash
        continue-on-error: ${{ fromJSON(env.CONTINUE_ON_ERROR) }}
```

**Comment by mike-carey** [comment]
Note that even though boolean can be "cast" to string, string cannot be "cast" to boolean and results in internal error.

Reusable workflow

`.github/workflows/accepts-boolean.yml`:

```yaml
---

name: 'Accepts boolean input'

on:
  workflow_call:
    inputs:
      foo:
        type: boolean
        default: false

jobs:
  test:
    name: 'Test'
    runs-on: ubuntu-latest
    steps:
      - run: echo ${{ inputs.foo }}
        shell: bash
```

Calling workflow:

```yaml
---

name: 'Uses a reusable workflow with boolean input'

on: [workflow_dispatch]

jobs:
  info:
    name: 'Gather information'
    runs-on: ubuntu-latest
    outputs:
      my-boolean-output: ${{ steps.info.outputs.my-boolean-output }}
    steps:
      - run: echo "::set-output name=my-boolean-output::true"
        shell: bash
        id: info

  success:
    name: 'Properly "casts" the string to boolean'
    needs: info
    uses: ./.github/workflows/accepts-boolean.yml
    with:
      foo: ${{ needs.info.outputs.my-boolean-output == 'true' }}
  
  failure:
    name: 'Incorrectly "casts" the string to boolean'
    needs: info
    uses: ./.github/workflows/accepts-boolean.yml
    with:
      foo: ${{ needs.info.outputs.my-boolean-output }}
```

<img width="1063" alt="image" src="https://user-images.githubusercontent.com/32496966/153673775-9df4de13-d0d2-427d-880a-091a4517860e.png">

References:
  - [Reusable workflow: accepts-boolean.yml](https://github.com/mike-carey/test_repo/blob/main/.github/workflows/accepts-boolean.yml)
  - [Caller workflow: boolean-error.yml](https://github.com/mike-carey/test_repo/blob/main/.github/workflows/boolean-error.yml)
  - [Run](https://github.com/mike-carey/test_repo/actions/runs/1831638921)

**Comment by kagahd** [comment]
I updated my comment above because the input variable type has to be `string` instead of `boolean` which is  casted to `boolean` by `fromJSON` later on when the variable is being read and used.

**Comment by mike-carey** [comment]
It looks like a more descriptive error is being echoed when passing a string as a boolean has been added.

<img width="1887" alt="image" src="https://user-images.githubusercontent.com/32496966/154369029-e08d8bdd-69b0-4531-9691-4498657bb79a.png">

I am going to hide my comment as it has been "resolved" and is no longer relevant since the error message will help decode the issue.

**Comment by mike-carey** [comment]
Something that I also recently learned is that booleans are "strings" within workflow dispatches and composite actions but not reusable workflows.

For workflow_dispatch and composite action boolean inputs are actually strings:

```yaml
if: inputs.my-boolean == 'true'
```

and

```yaml
if: github.event.inputs.my-boolean == 'true'
```

will work.

However, for reusable workflows, the boolean is actually a boolean:

```yaml
if: inputs.my-boolean == true
```

or

```yaml
if: inputs.my-boolean
```

## References
- [Demo workflow using dispatch](https://github.com/mike-carey/test_repo/blob/demo/boolean-inputs/.github/workflows/check-booleans.yml)
- [Demo composite action](https://github.com/mike-carey/test_repo/blob/demo/boolean-inputs/.github/actions/boolean-input/action.yml)
- [Demo reusable workflow](https://github.com/mike-carey/test_repo/blob/demo/boolean-inputs/.github/workflows/reusable-boolean-input.yml)
- [Demo run](https://github.com/mike-carey/test_repo/actions/runs/1855665225)

![image](https://user-images.githubusercontent.com/32496966/154371861-c2fd32d0-760d-420a-a9fe-2306638c323c.png)
![image](https://user-images.githubusercontent.com/32496966/154371878-863da5b2-b698-46df-b974-c7e9dc141aa1.png)
![image](https://user-images.githubusercontent.com/32496966/154371911-332d823b-385a-4ee9-9440-e1219a95f228.png)

**Comment by allenoooo** [comment]
This issue caused me to test booleans in actions, and it seems that setting a boolean behaves as expected, and is not a string, but a true boolean (heh). 

Can someone confirm? perhaps this issue should be closed.

**Comment by ryanwholey** [comment]
This is how boolean values are evaluated across dispatched inputs vs callable workflow inputs:

ex:
```yaml
// callable workflow
  workflow_call:
    inputs:
      boolean:
        type: boolean
        required: false
...
// dispatched workflow
  workflow_dispatch:
    inputs:
      boolean:
        required: false
        type: boolean
```

|| `== null` | `== false` | `== true` | `== 'false'` | `== 'true'`|
|---|---|---|---|---|---|
|dispatched bool false | f | f | f | **true** | f |
|dispatched bool true | f | f | f | f | **true** |
|workflow input unset | **true** | **true** | f | f | f |
|workflow input false | **true** | **true** | f | f | f |
|workflow input true | f | f | **true** | f | f |

**Comment by ajaykn** [comment]
👋 We are working on some efforts in unifying the inputs context and will address these issues in future.
So it should behave like inputs.boolean == true instead of strings ("true")
We will update once rolled out.

**Comment by Im2451soon** [comment]
멎있어요

**Comment by Im2451soon** [comment]
굿

**Comment by Im2451soon** [comment]
👍 👍 👍 👍 👍 👍 👍

**Comment by scottwn** [comment]
@ryanwholey's [comment](https://github.com/actions/runner/issues/1483#issuecomment-1090972149) above is not consistent with the [documented](https://github.com/github/docs/pull/16028) [behavior](https://docs.github.com/en/actions/learn-github-actions/contexts#example-usage-of-the-inputs-context)

**Comment by hex0cter** [comment]
+1

**Comment by ajaykn** [comment]
> 👋 We are working on some efforts in unifying the inputs context and will address these issues in future. So it should behave like inputs.boolean == true instead of strings ("true") We will update once rolled out.

Hey everyone, 
we have recently released some improvements.
Blog post: https://github.blog/changelog/2022-06-10-github-actions-inputs-unified-across-manual-and-reusable-workflows/

Now `workflow_dispatch` inputs can be accessed via `inputs` context directly and respecting the boolean type.
Please have a look and get back to us for any queries or any usecases. cc @chrispat @avocaddo

**Comment by maratori** [comment]
When I run job manually boolean input still processed as string and following config always runs tmate.
```yaml
on:
  workflow_dispatch:
    inputs:
      debug_enabled:
        description: "Run the build with tmate debugging enabled (https://github.com/marketplace/actions/debugging-with-tmate)"
        required: false
        type: boolean
        default: false

jobs:
  ci:
    runs-on: ubuntu-latest
    steps:
      - name: Setup tmate session
        uses: mxschmitt/action-tmate@v3
        if: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.debug_enabled }}
```

**Comment by ajaykn** [comment]
I think we can close this issue :) cc @thboop

**Comment by nathanchapman** [comment]
@mikewu-tl booleans have historically worked as expected with `workflow_call` inputs, used via `inputs.<input_name>`.
The code changes you'd need to make on your end are switching from `github.event.inputs.<input_name>` to just `inputs.<input_name>`. Another advantage here is this unifies how you'd access inputs for workflows that are both reusable (`workflow_call`) and manually triggered (`workflow_dispatch`). The ability to do this was just released last Friday, June 10th, 2022 and it technically solves two issues.

**Comment by ajaykn** [comment]
@mikewu-tl Sorry if i dint make it clear in my [comment](https://github.com/actions/runner/issues/1483#issuecomment-1154811601).
As @nathanchapman explained it well, we need to update the workflows with `inputs.boolean-foo` instead of `github.event.inputs.boolean-foo`.
We dint go ahead and fix the boolean types in github context, because few customers are already using `githib.event.inputs.boolean-foo == 'false'` and we dint want to break those existing ones.

Wanted to explain again what scenarios we are covering with unifying improvements:
### Scenario 1: 
**Simplify the workflow for both reusable and manually triggered**
Some workflow authors who wanted to have one workflow that was both reusable and manually triggered can be achieved and can use the inputs context to access the input values.

Example-1, if we want to set USERNAME, then customers have to go through this sort of code.
```yml
WORKFLOW_DISPATCH_INPUT=${{ github.event.inputs.username }}
WORKFLOW_CALL_INPUT=${{ inputs.username }}
USERNAME=${WORKFLOW_DISPATCH_INPUT:-$WORKFLOW_CALL_INPUT}
```
Now, we can simplify with `USERNAME=${{ $inputs.username }}`

Example-2: Customer used to write a wrapper workflow to manually trigger a reusable one
```yml
name: Dispatching reusable workflow
on:
  workflow_dispatch:
    inputs:
      key1:
        default: value1
jobs:
  job1:
    uses: org/repo/.github/workflows/reusable-workflow.yml@main
    with:
      key1: ${{ github.event.inputs.key1 }}
```
Now customer can get rid of this wrapper workflow.


### Scenario 2:
Respect boolean types for workflow_dispatch inputs.
Previously customers used to write `githib.event.inputs.boolean-foo == 'false'`, now it can be replaced with `inputs.boolean-foo` directly in workflow.

**Comment by thboop** [comment]
[This should be fixed now](https://github.blog/changelog/2022-06-10-github-actions-inputs-unified-across-manual-and-reusable-workflows/) for workflow_dispatch, please let us know if you see any issues

**Comment by caffeineflo** [comment]
I think I found another issue unfortunately.

Consider the following reusable workflow:

```
name: Test
on:
  workflow_dispatch:
    inputs:
      debug_enabled:
        description: 'Run the build with debugging enabled?'
        required: false
        default: false
        type: boolean

jobs:
  print_input:
    runs-on: ubuntu-latest
    steps:
      - run: echo "Input debug enabled is:"
      - run: echo ${{ inputs.debug_enabled }}

  another_step:
    uses: /PATH/.github/workflows/another_step.yaml@main
    with:
      debug_enabled: ${{ inputs.debug_enabled }}
```


Now the `print_input` step prints "Input debug is enabled: true", but `another_step` is never triggered and I get an annotation saying:
`Unexpected type of value '', expected type: Boolean` for the last line which should have evaluated to a true boolean.

Please correct me if I'm wrong, but either the variable expansion doesn't work or the input is still the wrong type.

Thanks!

**Comment by nathanchapman** [comment]
@caffeineflo I ran into that as well, this did the trick for me:

```
  another_step:
    uses: /PATH/.github/workflows/another_step.yaml@main
    with:
      debug_enabled: ${{ inputs.debug_enabled == true }}
```

**Comment by ChristopherHX** [comment]
This is pretty shure a variable expansion problem, but I didn't even know that ` jobs.<job_id>.with.<with_id>` allows the `inputs` context. It isn't even documented here https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability.

> another_step is never triggered and I get an annotation saying:
`Unexpected type of value '', expected type: Boolean` for the last line which should have evaluated to a true boolean.

The `inputs` context is null and expanding a property of a null object leads to null, null is not a boolean and you see an error. I guess the bug is a not populated inputs context of the expression evaluator.

You can see that inputs is null by using tojson
```yaml
on:
  workflow_dispatch:
    inputs:
      test:
        type: choice
        options:
        - mychoice
        - a
        - b
      b:
        type: boolean
jobs:
  _:
    uses: ./.github/workflows/call.yml
    with:
      u: ${{ tojson(inputs) }} # u is assigned to string 'null'
```

**Comment by nathanchapman** [comment]
@ChristopherHX

> This is pretty shure a variable expansion problem, but I didn't even know that jobs.<job_id>.with.<with_id> allows the inputs context. It isn't even documented here https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability.

This was part of the [inputs unification announcement](https://github.blog/changelog/2022-06-10-github-actions-inputs-unified-across-manual-and-reusable-workflows/) on June 10th.

The behavior you're experiencing is a separate bug, detailed here: https://github.com/actions/runner/issues/1976

It appears to be a regression introduced sometime in the last week. I believe in [v2.294.0](https://github.com/actions/runner/releases/tag/v2.294.0)

**Comment by caffeineflo** [comment]
> @caffeineflo I ran into that as well, this did the trick for me:
> 
> ```
>   another_step:
>     uses: /PATH/.github/workflows/another_step.yaml@main
>     with:
>       debug_enabled: ${{ inputs.debug_enabled == true }}
> ```
> 
> Note: this is only required when calling a reusable workflow that takes a boolean as input. Using `if: ${{ inputs.debug_enabled }}` on jobs or steps within the same workflow file works correctly
> 
> I agree with you on it likely being an issue with variable expansion

Thanks! That worked and saved my night

**Comment by ajaykn** [comment]
@caffeineflo Please find the latest update here
https://github.com/actions/runner/issues/1976#issuecomment-1172940227

tldr:
You should be able to pass directly `debug_enabled: ${{ inputs.debug_enabled }}` and should work fine.

**Comment by nathanchapman** [comment]
@ajaykn I believe there's still an issue here when the dispatchable workflow is also triggered by an event (like `github.event.pull_request`). Since the event won't have `debug_enabled` set, and there's no way to provide a workflow-level default, a type error will be thrown when a pull request triggers the workflow, even though it'll work fine for a manual dispatch. This happens because the re-usable workflow expects a boolean and receives a null value.

**Comment by ventsyv** [comment]
I've ran into this as well and `if: ${{ inputs.x == 'true' }}` doesn't work. Here is my setup:

`name: Test Build

on:
  push:
    branches:
      - main

  pull_request:
    branches:
      - main

  workflow_dispatch:
    inputs:
      test:
        description: 'Test boolean'
        required: true
        type: boolean
        default: true

jobs:
  test_job:
    runs-on: self-hosted
    container:
      image: ubuntu:latest

    steps:
      - uses: actions/checkout@v2
      
      - name: Some step
        run: echo "test"
      
      - name: Test step
          if: ${{ inputs.test == 'true' }}
          run: echo "Flag is TRUE"`

This is on GitHub Enterprise Server 3.5.3. Reading through the thread, I was under the impression that using string comparison should work? What gives?

**Comment by nathanchapman** [comment]
@ventsyv I believe you should be using `if: ${{ inputs.x == true }}`. A lot has been fixed here, but IIRC, it was `github.events.inputs.x` where strings were used for booleans

**Comment by ventsyv** [comment]
That did not work.
I thought the root of the problem is that the inputs are not true booleans but actually strings, thus the `inputs.x == 'true'`

**Comment by nathanchapman** [comment]
That used to be true but is no longer the case, which is why this issue was closed. I can confirm it's been resolved, at least for Free/Pro/Teams and Enterprise Cloud. If you're still experiencing that on Enterprise Server, the best course of action may be to raise a support ticket:
https://github.com/premium-support

**Comment by rainabba** [comment]
I am also on Enterprise and also finding "boolean" params unreliable with these `Unexpected type of value 'true', expected type: Boolean` errors despite my param being declared `type: boolean` and using `var == true` or just `var`

My workaround is use `${{needs.defaults.outputs.myoutput == true` rather than relying on myoutput to "be a boolean".

Here's a full test-case that represents the issue though you'll have to update the actual workflow referenced to a real one to test.

```
name: Sample boolean handling bug

on:
  pull_request:
    branches: [ main,develop,develop ]

  workflow_dispatch:
    inputs:
      BOOLEAN_VAR:
        required: true
        type: boolean
        default: true

jobs:
  defaults:
    runs-on: [self-hosted,common-2gb]
    env:
      BOOLEAN_VAR: ${{ inputs.BOOLEAN_VAR != '' && inputs.BOOLEAN_VAR || true }}
    steps:
      - name: Process inputs and defaults
        id: defaults
        run: |
          echo "::set-output name=BOOLEAN_VAR::${{ env.BOOLEAN_VAR  }}";
    outputs:
      BOOLEAN_VAR: ${{ steps.defaults.outputs.BOOLEAN_VAR}}

  failing-job:
    needs: defaults
    uses: enterprise-org/some-repo/.github/workflows/pr-merge-checks.yml@pr-merge-checks-v1
    with:
      BOOLEAN_VAR: ${{ needs.defaults.outputs.BOOLEAN_VAR }}
      # Errors out with `Unexpected type of value 'true', expected type: Boolean`

  suceeding-job:
    needs: defaults
    uses: enterprise-org/some-repo/.github/workflows/pr-merge-checks.yml@pr-merge-checks-v1
    with:
      BOOLEAN_VAR: ${{ needs.defaults.outputs.BOOLEAN_VAR == true }}
      # This one will succeed
```

**Comment by scott-doyland-burrows** [comment]
I find that if I use the following input:

```
inputs:
  myar:
    required: false
    type: boolean
    default: false
```

If set in a composite action then ```myvar``` is a string, and accepts string input.
If I set in a reusable workflow then ```myvar``` is a boolean.

**Comment by flobernd** [comment]
@MetRonnie This is not fixed for composite actions!

I have a simple action that just wraps the `create-release` API:

```yaml
inputs:
  ...
  generate-release-notes:
    description: ...
    required: false
    type: boolean
    default: false

runs:
  using: composite
  steps:
    - name: Create Release
      uses: actions/github-script@v6
      with:
        script: |
          github.rest.repos.createRelease({
            owner: context.repo.owner,
            repo: context.repo.repo,
            ...
            generate_release_notes: ${{ inputs.generate-release-notes && 'true' || 'false' }}
          });
```

Caller:

```yaml
- uses: flobernd/actions/github/create-release@master
  with:
    tag-name: v1.2.4
    generate-release-notes: true
```

This line always evaluates to `false`:

```
generate_release_notes: ${{ inputs.generate-release-notes && 'true' || 'false' }}
```

The explicit syntax does incorreclty evaluate to `false` as well:

```
generate_release_notes: ${{ inputs.generate-release-notes == true && 'true' || 'false' }}
```

Correct behavior is only observed when using string semantics:

```
generate_release_notes: ${{ inputs.generate-release-notes == 'true' && 'true' || 'false' }}
```

**Comment by ajaykn** [comment]
@flobernd Apologies for replying late.
I am not sure if we support `type` for actions inputs.
https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_id

We support for workfow_dispatch inputs, workflow call inputs though
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_dispatchinputs
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

And this issue talks about workflow_dispatch.

> @MetRonnie @TingluoHuang @ericsciple This is not fixed for composite actions!
>
> I have a simple action that just wraps the create-release API:
    
So i think for composite actions, since we don't support inputs type, we are considering them as string.

We will track a different issue as feature enhancement separately :)

**Comment by dekimsey** [comment]
> We will track a different issue as feature enhancement separately :)

Hi @ajaykn, which issue will be used to track this behavior? #2238? It looks to be this same issue, but I'm not sure. Would you clarify? I'd like to make sure I'm subscribed to the right ticket and also able to escalate this to our GH liason.

Thank you!

**Comment by DarrenBishop** [comment]
Having similar issues, _variation on a theme_:

```
name: Build Images

on:
  push:
    ...
  pull_request:
    ...

permissions:
  ...

jobs:
  determine-work:
     ...
  call-build-image:
    needs: determine-work
    strategy:
      matrix:
        image: [api, ui]
        include:
          - image: api
            enabled: ${{ needs.determine-work.outputs.build-api }}
          - image: ui
            enabled: ${{ needs.determine-work.outputs.build-ui }}
    uses: org/repo/.github/workflows/build_image.yaml@same-branch
    with:
      build: ${{ matrix.enabled }}
      #build: ${{ fromJSON(matrix.enabled) }} # <--- fromJSON FTW
      image: ${{ matrix.image }}
      tag: ${{ needs.determine-work.outputs.tag }}
    secrets: inherit
    permissions:
      id-token: write
      contents: read
```

...and the called workflow

```
name: Build Single Image

on:
  workflow_call:
    inputs:
      build:
        description: 'Flag whether to build the Image'
        required: true
        type: boolean
      image:
        description: 'Image to build: api or ui'
        required: true
        type: string
      tag:
        description: 'Image tag'
        required: true
        type: string

jobs:
  build-and-push:
    if: inputs.build
    ...
```

Seems the `matrix` involvement undermines all the above e.g. reverting/coercing all things to strings; I was getting two annotations with the same report:
```
The template is not valid. (Line: 187, Col: 14): Unexpected value 'true'
```
Note no mention of the expected `boolean` type.

Also note the annoying hack to workaround the `matrix` context not being available for use in `jobs.<job_id>.if`

The _`fromJSON`_ trick mentioned above (credit to the appropriate community member) did the trick... it only took 50 odd failed runs of debugging to get here.

**Comment by markwhitfeld** [comment]
To convert the "boolean" input of a `workflow_dispatch` to be consumable by the boolean input of a `workflow_call` I used the following logic:
`${{ input.my-bool == true && input.my-bool == 'true' }}`
This will accept a boolean `true` or the string `"true"` as input to make it true, anything else is false.

**Comment by piotrekkr** [comment]
> To convert the "boolean" input of a `workflow_dispatch` to be consumable by the boolean input of a `workflow_call` I used the following logic: `${{ input.my-bool == true && input.my-bool == 'true' }}` This will accept a boolean `true` or the string `"true"` as input to make it true, anything else is false.

@markwhitfeld You mean `${{ input.my-bool == true || input.my-bool == 'true' }}`? Otherwise a variable cannot be `bool` and `string` at same time so at least one of conditions will always fail and whole expression will be `false`.

**Comment by davidtchow** [comment]
Also see https://github.com/actions/runner/issues/1483

**Comment by Uzlopak** [comment]
@markwhitfeld 
Shouldnt it be 
`${{ input.my-bool == true || input.my-bool == 'true' }}`

?

**Comment by Kristonitas** [comment]
This isn't happening any more since some time 2023: https://github.com/github/docs/commit/e7a6d023cee78ddc4e5ac813cd0d1cb8eaac923d

Inputs are still strings (when using `gh workflow run`, but the "boolean" input behaves like an enum with `"true"` and `"false"` being the only two options that are later cast to a proper boolean. I imagine numbers could be working same way, but didn't validate that.

I'm confused about the recent activity referencing this issue. Did the problem reappear at some point?
![image](https://github.com/actions/runner/assets/9002285/e0c826aa-6911-47e6-9a20-ebe69775a885)

**Comment by libermansa** [comment]
well, this is still not working as expected. putting if: ${{ inputs.SOME_BOOLEAN }} doesn't evaluate booleans.

**Comment by piotrekkr** [comment]
> This isn't happening any more since some time 2023: [github/docs@e7a6d02](https://github.com/github/docs/commit/e7a6d023cee78ddc4e5ac813cd0d1cb8eaac923d)
> 
> Inputs are still strings (when using `gh workflow run`, but the "boolean" input behaves like an enum with `"true"` and `"false"` being the only two options that are later cast to a proper boolean. I imagine numbers could be working same way, but didn't validate that.
> 
> I'm confused about the recent activity referencing this issue. Did the problem reappear at some point? ![image](https://private-user-images.githubusercontent.com/9002285/334794907-e0c826aa-6911-47e6-9a20-ebe69775a885.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgwMDUwODAsIm5iZiI6MTcxODAwNDc4MCwicGF0aCI6Ii85MDAyMjg1LzMzNDc5NDkwNy1lMGM4MjZhYS02OTExLTQ3ZTYtOWEyMC1lYmU2OTc3NWE4ODUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTBUMDczMzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjViNThjNjJjMjcyMzYyMTE3YjNiNzVjYjg1YTA2ZDM5YWUwOWEyZTFmMWEyOGY5N2NiZTBmMGZlOGFmNDM4YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.aJHpYM1iL5-NPClco_8TTRtfHhg2MxgNzIuOvzgJA7I)

@Kristonitas I think those issues you added in screenshot were about output params and not inputs. Maybe outputs were mixed with inputs by some people and those were incorrectly linked to this issue.

**Comment by Kristonitas** [comment]
@libermansa 
I remade the test that was posted early in this issue, with results that we expect from bool inputs:
https://github.com/Kristonitas/ga-bool-input-test
<img width="317" alt="image" src="https://github.com/actions/runner/assets/9002285/2067f80b-d934-4c61-88a0-7555c0ac8d88">
<img width="338" alt="image" src="https://github.com/actions/runner/assets/9002285/d40dad59-348d-4db6-a462-581d493ea25a">

The scenario that was reported in https://github.com/actions/runner/issues/1483#issuecomment-1090972149 (and other similar tests) is not happening right now

@piotrekkr 
Thanks for pointing that out

P. S. The arguments passed between workflows are still serialised as strings afaik (hence you can't use json booleans when using the json input flag `echo '{"passed_true": true, "passed_false": false}' | gh workflow run test.yml --json` would fail). So seeing strings as output "booleans" could still make sense.

**Comment by rainabba** [comment]
I found that I needed to use `fromJSON()` (see: https://docs.github.com/en/actions/learn-github-actions/expressions#fromjson) around my "boolean" params to get them to behave as expected.

**Comment by sanjacob** [comment]
Still happens in env at least. I had to use `fromJSON`

```
inputs:
  dev:
    description: "Also add development requirements"
    default: false
    type: boolean

steps:
    - env:
        ARG_DEV: ${{ fromJSON(inputs.dev) && '--dev' || '' }}
      run:  pipenv install $ARG_DEV
      shell: bash
```
